Implement Partition V3 Kafka REST API

rohitsanj opened the pull request:
<!-- Consider adding [ci skip] to the PR title if the PR does not change the source code or tests. -->

## Summary of Changes

<!-- Include a high-level overview of your implementation, including any alternatives you considered and items you'll address in follow-up PRs -->

Followup item from #74

- Implemented get and list topic partitions
- I've refactored the `TopicManagerImpl` to make its methods return plain Kafka objects rather than directly constructing API models, which is now done in `TopicV3ApiImpl`
- Set `client.id` config for AdminClient to `Confluent for VS Code sidecar ${quarkus.application.version}` (discussed as followup in #74)

## Any additional details or context that should be provided?

<!-- Behavior before/after, more technical details/screenshots, follow-on work that should be expected, links to discussions or issues, etc -->


## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [x] Added new
    - [x] Updated existing
    - [ ] Deleted existing
- [x] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [x] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

rohitsanj commented:
I tested out the native executable built from this branch end-to-end from the VS Code extension -- as shown in screenshot, the message viewer tab is able to list topic partitions.

<img width="1661" alt="Screenshot 2024-10-10 at 2 15 23 PM" src="https://github.com/user-attachments/assets/4257368d-52e7-4030-beda-41a526ee6aaa">

rohitsanj commented:
@rhauch For reference, this is what kafka-rest constructs the CRN as for a topic partition:

```
 "resource_name": "crn:///kafka=4L6g3nShT-eMCtK--X86sw/topic=test2/partition=0"
```

<details>
<summary> Full sample list partitions response </summary>

```
{
  "kind": "KafkaPartitionList",
  "metadata": {
    "self": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions",
    "next": null
  },
  "data": [
    {
      "kind": "KafkaPartition",
      "metadata": {
        "self": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/0",
        "resource_name": "crn:///kafka=4L6g3nShT-eMCtK--X86sw/topic=test2/partition=0"
      },
      "cluster_id": "4L6g3nShT-eMCtK--X86sw",
      "topic_name": "test2",
      "partition_id": 0,
      "leader": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/0/replicas/1"
      },
      "replicas": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/0/replicas"
      },
      "reassignment": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/0/reassignment"
      }
    },
    {
      "kind": "KafkaPartition",
      "metadata": {
        "self": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/1",
        "resource_name": "crn:///kafka=4L6g3nShT-eMCtK--X86sw/topic=test2/partition=1"
      },
      "cluster_id": "4L6g3nShT-eMCtK--X86sw",
      "topic_name": "test2",
      "partition_id": 1,
      "leader": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/1/replicas/1"
      },
      "replicas": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/1/replicas"
      },
      "reassignment": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/1/reassignment"
      }
    },
    {
      "kind": "KafkaPartition",
      "metadata": {
        "self": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/2",
        "resource_name": "crn:///kafka=4L6g3nShT-eMCtK--X86sw/topic=test2/partition=2"
      },
      "cluster_id": "4L6g3nShT-eMCtK--X86sw",
      "topic_name": "test2",
      "partition_id": 2,
      "leader": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/2/replicas/1"
      },
      "replicas": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/2/replicas"
      },
      "reassignment": {
        "related": "http://rest-proxy:8082/v3/clusters/4L6g3nShT-eMCtK--X86sw/topics/test2/partitions/2/reassignment"
      }
    }
  ]
}
```

</details>

flippingbits commented on src/main/java/io/confluent/idesidecar/restapi/kafkarest/Manager.java:
```diff
@@ -0,0 +1,28 @@
+ * be annotated with {@link jakarta.enterprise.context.RequestScoped} to allow injection of
+ * {@link HttpServerRequest}.
+ */
+public abstract class Manager {
+  @Inject
+  AdminClients adminClients;
+
+  @Inject
+  ClusterCache clusterCache;
+
+  @Inject
+  HttpServerRequest request;
+
+  Supplier<String> connectionId = () -> request.getHeader(CONNECTION_ID_HEADER);
+}
```
I'd probably remove this abstract base class because it does not hold any implementation and move the `@Inject`ed dependencies to _manager_ classes that need them.

flippingbits commented on src/main/java/io/confluent/idesidecar/restapi/kafkarest/Manager.java:
```diff
@@ -0,0 +1,28 @@
+  AdminClients adminClients;
+
+  @Inject
+  ClusterCache clusterCache;
```
Neither the `PartitionManagerImpl` nor the `TopicManagerImpl` seem to use the `clusterCache`.

flippingbits commented on src/main/java/io/confluent/idesidecar/restapi/kafkarest/PartitionManagerImpl.java:
```diff
@@ -0,0 +1,39 @@
+@RequestScoped
+public class PartitionManagerImpl extends Manager {
+  @Inject
+  TopicManagerImpl topicManager;
```
```suggestion
  TopicManager topicManager;
```

Do you know if we can inject the interface, `TopicManager`, instead of the class, `TopicManagerImpl`, here? In general, interfaces ensure that components are loosely coupled and might ease mocking in tests.

rohitsanj replied:
Yes, but I remember encountering some sort of a bean injection issue that's not occurring anymore, maybe I was doing something else wrong. I've made changes across kafkarest classes to inject interfaces wherever possible rather than implementations.

flippingbits commented on src/main/java/io/confluent/idesidecar/restapi/kafkarest/TopicV3ApiImpl.java:
```diff
@@ -31,12 +46,30 @@ public Uni<Void> deleteKafkaTopic(String clusterId, String topicName) {
   @Override
   public Uni<TopicDataList> listKafkaTopics(String clusterId, Boolean includeAuthorizedOperations) {
-    return topicManager.listKafkaTopics(clusterId, includeAuthorizedOperations);
+    return topicManager.listKafkaTopics(clusterId, includeAuthorizedOperations)
+            .onItem().transformToUni(topicDescriptionMap -> uniItem(TopicDataList
+            .builder()
+            .kind("KafkaTopicList")
+            .metadata(ResourceCollectionMetadata
+                .builder()
+                .next(null)
+                .self(forTopics(clusterId).getRelated())
+                .build()
+            )
+            .data(topicDescriptionMap
+                .stream()
+                .map(t -> fromTopicDescription(clusterId, t))
+                .toList()
+            ).build())
+        );
```
Might be easier to read:

```suggestion
    return topicManager
        .listKafkaTopics(clusterId, includeAuthorizedOperations)
        .onItem()
        .transformToUni(topicDescriptionMap ->
            uniItem(
                TopicDataList
                  .builder()
                  .kind("KafkaTopicList")
                  .metadata(
                      ResourceCollectionMetadata
                        .builder()
                        .next(null)
                        .self(forTopics(clusterId).getRelated())
                        .build()
                  )
                  .data(
                      topicDescriptionMap
                        .stream()
                        .map(topic -> fromTopicDescription(clusterId, topic))
                        .toList()
                  )
                  .build()
            )
        );
```

flippingbits reviewed (commented):
Thanks, @rohitsanj! I have a few suggestions. Otherwise, your PR looks good to me.

rohitsanj commented:
@flippingbits: thanks for the review, I've addressed the feedback. Note that I've also added a `PartitionManager` interface class that I'd missed adding initially. Would you be able to take another look?
